src/devices/src/legacy/acpi_device.rs

```diff
@@ -0,0 +1,166 @@
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// | 0 ... 7       8        9 ... 19       24      25 ... 31 | 32  ...  41      42          43          44         45       46 47 | 48 .. 64 |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// |                        X_PM1a_EVT_BLK                   |                                  X_PM1a_CNT_BLK                    | NOT USED |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// | 0 ... 7       8        9 ... 19       24      25 ... 31 |  0  ...   9      10          11          12         13       14 15 | NOT USED |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// |           PWRBTN_STS              PWRBTN_EN             |               SLP_TYPx    SLP_TYPx    SLP_TYPx    SLP_EN           | NOT USED |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+
+// To trigger a shutdown, PWRBTN_STS and PWRBTN_EN (in X_PM1a_EVT_BLK) needs to be written as
+// ones and the interrupt should be raised. When guest goes to soft-off state, SLP_EN is written
+// as one and SLP_TYPx is written with the value from the _S5 object from the DSDT table (in
+// fact, when guest goes into power state x, then the value from the _Sx object is written into
+// SLP_TYPx bits located into X_PM1a_CNT_BLK)
+
+pub struct AcpiDevice {
+    /// Data Register
+    data: u64,
+    /// Interrupt to be triggered for OSPM to read ACPI registers
+    interrupt: EventFd,
+    /// CPU reset eventfd. We will set this event to stop Firecracker.
+    reset_evt: EventFd,
+}
+
+impl AcpiDevice {
+    /// Constructs a device that holds ACPI registers.
+    pub fn new(interrupt: EventFd, reset_evt: EventFd) -> Self {
+        Self {
+            data: 0,
+            interrupt,
+            reset_evt,
+        }
+    }
+
+    pub fn send_acpi_shutdown_signal(&mut self) -> Result<()> {
+        const PWRBTN_STS_OFFSET: u64 = 8;
+        const PWRBTN_EN: u64 = 24;
+
+        // Set bits according to the spec in order to trigger the shutdown
+        // https://uefi.org/specs/ACPI/6.4/04_ACPI_Hardware_Specification/ACPI_Hardware_Specification.html#pm1-event-grouping
+        self.data = self.data | (1 << PWRBTN_STS_OFFSET) | (1 << PWRBTN_EN);
+
+        // Make OSPM read ACPI registers and generate the ACPI event by setting the interrupt on ACPI Device
+        self.interrupt
+            .write(9) // SCI_INT value from ACPI FADT table
```

Comment:
the idiomatic way to trigger an eventfd is `write(1)` which can be easily mentally matched to `write(true)`.
While `write(9)` also works, it is confusing and can be misconstrued as sending 9 separate interrupts.

The irq number (9) for the raised IRQ tied to this eventfd is set during `register_irqfd()` and is not related to the amount written to the eventfd. That's why `write(1)` is enough, it says this eventfd got 1 event => Guest IRQ `#9` tied to it also gets 1 event.

(Also, `ACPI_SCI_INT` constant is defined in acpi crate, better to use constants than hardcode numbers)

Reply:
It makes sense to be so... At the beginning, it was a bit unclear for me how interrupt triggering works ... and then I discovered register_irqfd() function that did the mapping between the irq number and eventfd. But I also kept write(9) thinking that it is still required, but the mapping was already done in register_irqfd(). Thank you so much for the explanation!